_posts/you-dont-need-that-abstraction.md

```diff
@@ -0,0 +1,57 @@
+
+Don't get me wrong; testing is a good thing, crucial, I'd say. However, here the problem is that integration testing is arduous or even impossible due to the monolithic nature of the system. That leads the design to need abstraction layers.
+
+[Derek Comartin](https://codeopinion.com/) puts it very nicely in ["What's the cost of indirection & abstraction?"](https://codeopinion.com/whats-the-cost-of-indirection-abstractions/):
+
+> The benefits are reuse, isolating complexity, encapsulation of dependencies, and more. But what's the cost of indirection & abstractions? Cognitive load to fully understand all of the layers of a request and limiting functionality.
+
+## The SOA way
+
+A service-oriented architecture can have a dramatic impact on the way we approach all those abstraction layers. Again, don't get me wrong, it's possible to design an utterly flawed SOA-based system, something usually referred to as a distributed monolith. However, SOA removes some of the barriers mentioned above that lead to the need for abstraction layers.
+
+Autonomous components are one of the SOA building blocks. They are the essence of the single responsibility principle in a distributed system. Autonomy comes with significant advantages, one of which is isolation. A component that is both autonomous and isolated is also independent. Independence means that we're free to host it alone.
+
+If a component has all the mentioned attributes, it can be easily tested, which is true in a monolithic architecture. What's different is the hosting aspect. If we can host a component alone in its process, [integration testing](https://milestone.topics.it/2021/04/07/nservicebus-integrationtesting-baby-steps.html) comes at ease.
+
+If integration testing is one of the tools in our toolbox, there is less need for abstraction layers. An autonomous component whose responsibility is to deal, for example, with loyalty points in an e-commerce system, can directly access its storage without an abstraction layer. We can quickly test the composition with the storage as a dependency. That is especially true nowadays, thanks to containers. It's never been so easy to spin up a dependency, like SQL Server or PostgreSQL, like it is today.
+
+Also, autonomous components drive more adherence to the single responsibility principle, and [NServiceBus message handlers](https://docs.particular.net/nservicebus/handlers/) are an excellent example. It means that the component can do one thing at a time. There is no simple workaround. It's much easier and riskier to add a method to a class in a monolithic architecture rather than adding additional behaviors to an autonomous component in a distributed system. There is so much more friction that we don't even try.
+
+Finally, code simplicity comes with less cognitive load.
+
+However, distributed systems come with their own set of complexity. Monitoring and debugging are representative samples of things that are harder than in monolithic architecture.
```

Comment:
This is a bit abrupt. You spend the bulk of the post debunking the myth that SOA is complicated, then admit that certain aspects of it are hard without commenting on it further. Is it so hard that it shifts the equation significantly? Are you going to expand on it in a future post?

Reply:
Good point, reworded.